src/shape.rs

```diff
@@ -348,3 +376,356 @@ mod test
+
+/// A struct representing an axis-aligned rounded rectangle. Two points and a float are
+/// stored: the top left vertex, the bottom right vertex and the radius of the
+/// rounded corners.
+///
+/// Alias for a rectangle with f32 coordinates.
+pub type RoundRect = RoundedRectangle<f32>;
+
+/// A struct representing an axis-aligned rounded rectangle. Two points and a float are
+/// stored: the top left vertex, the bottom right vertex and the radius of the
+/// rounded corners.
+#[derive(Debug, PartialEq, Eq, Clone)]
+#[repr(C)]
+pub struct RoundedRectangle<T = f32>
+{
+    top_left: Vector2<T>,
+    bottom_right: Vector2<T>,
+    radius: T
+}
+
+impl<T> AsRef<RoundedRectangle<T>> for RoundedRectangle<T>
+{
+    fn as_ref(&self) -> &Self
+    {
+        self
+    }
+}
+
+impl<T> RoundedRectangle<T>
+{
+    /// Constructs a new `RoundedRectangle`. The top left vertex must be above
+    /// and to the left of the bottom right vertex. A negative radius will
+    /// be clamped to 0. A big radius (larger than half the width or height)
+    /// might produce unexpected behavior but it won't be checked.
+    #[inline]
+    pub const fn new(top_left: Vector2<T>, bottom_right: Vector2<T>, radius: T) -> Self
+    {
+        RoundedRectangle {
+            top_left,
+            bottom_right,
+            radius
+        }
+    }
+
+    /// Constructs a new `RoundedRectangle`. The top left vertex must be above
+    /// and to the left of the bottom right vertex. A negative radius will
+    /// be clamped to 0. A big radius (larger than half the width or height)
+    /// might produce unexpected behavior but it won't be checked.
+    #[inline]
+    pub fn from_tuples(top_left: (T, T), bottom_right: (T, T), radius: T) -> Self
+    {
+        RoundedRectangle {
+            top_left: Vector2::new(top_left.0, top_left.1),
+            bottom_right: Vector2::new(bottom_right.0, bottom_right.1),
+            radius
+        }
+    }
+
+    /// Constructs a new `RoundedRectangle` from a `Rectangle` and a radius.
+    /// A big radius (larger than half the width or height) might produce
+    /// unexpected behavior but it won't be checked.
+    #[inline]
+    pub fn from_rectangle(rectangle: Rectangle<T>, radius: T) -> Self
+    {
+        RoundedRectangle {
+            top_left: rectangle.top_left,
+            bottom_right: rectangle.bottom_right,
+            radius
+        }
+    }
+
+    /// Returns a reference to the top left vertex.
+    #[inline]
+    pub const fn top_left(&self) -> &Vector2<T>
+    {
+        &self.top_left
+    }
+
+    /// Returns a reference to the bottom right vertex.
+    #[inline]
+    pub const fn bottom_right(&self) -> &Vector2<T>
+    {
+        &self.bottom_right
+    }
+}
+
+impl<T: Copy> RoundedRectangle<T>
+{
+    /// Returns a vector representing the top right vertex.
+    #[inline]
+    pub fn top_right(&self) -> Vector2<T>
+    {
+        Vector2::new(self.bottom_right.x, self.top_left.y)
```

Comment:
If we store a `Rectangle` in the struct, we can just call `self.rectangle.top_right()`. Same below too :)